### PR TITLE
Snapshot Validator BLSPKIDEntries

### PR DESCRIPTION
**Pull request by lazynina**

### CODE REVIEW

**Comment by lazynina** [comment]
Current dependencies on/for this PR:
* feature/pos-fast-hotstuff-consensus
  * **PR #742** <a href="https://app.graphite.dev/github/pr/deso-protocol/core/742?utm_source=stack-comment-icon" target="_blank">Graphite</a> 
    * **PR #812** <a href="https://app.graphite.dev/github/pr/deso-protocol/core/812?utm_source=stack-comment-icon" target="_blank">Graphite</a> 
      * **PR #809** <a href="https://app.graphite.dev/github/pr/deso-protocol/core/809?utm_source=stack-comment-icon" target="_blank">Graphite</a> 
        * **PR #813** <a href="https://app.graphite.dev/github/pr/deso-protocol/core/813?utm_source=stack-comment-icon" target="_blank">Graphite</a>  👈

This [stack of pull requests](https://stacking.dev/?utm_source=stack-comment) is managed by [Graphite](https://app.graphite.dev/github/pr/deso-protocol/core/813?utm_source=stack-comment).

File: lib/pos_snapshot_entries.go [hunk 1]
```diff
@@ -621,6 +624,184 @@ func DBEnumerateAllCurrentValidators(handle *badger.DB, pkidsToSkip []*PKID) ([]
...
+}
+
+func DBKeyForSnapshotValidatorBLSPublicKeyPKIDPairEntry(blsPublicKeyPKIDPairEntry *BLSPublicKeyPKIDPairEntry, snapshotAtEpochNumber uint64) []byte {
+	key := append([]byte{}, Prefixes.PrefixSnapshotValidatorBLSPublicKeyPKIDPairEntry...)
+	key = append(key, EncodeUint64(snapshotAtEpochNumber)...)
+	key = append(key, blsPublicKeyPKIDPairEntry.BLSPublicKey.ToBytes()...)
+	return key
+}
+
+func DBGetSnapshotValidatorBLSPublicKeyPKIDPairEntryByBLSPublicKey(handle *badger.DB, snap *Snapshot, blsPublicKey *bls.PublicKey, snapshotAtEpochNumber uint64) (*BLSPublicKeyPKIDPairEntry, error) {
```
Comment (tholonious) [inline lib/pos_snapshot_entries.go:671]:
```suggestion
func DBGetSnapshotValidatorBLSPublicKeyPKIDPairEntry(handle *badger.DB, snap *Snapshot, blsPublicKey *bls.PublicKey, snapshotAtEpochNumber uint64) (*BLSPublicKeyPKIDPairEntry, error) {
```

File: lib/pos_snapshot_entries.go [hunk 2]
```diff
@@ -621,6 +624,184 @@ func DBEnumerateAllCurrentValidators(handle *badger.DB, pkidsToSkip []*PKID) ([]
...
+	var ret *BLSPublicKeyPKIDPairEntry
+	err := handle.View(func(txn *badger.Txn) error {
+		var innerErr error
+		ret, innerErr = DBGetSnapshotValidatorBLSPublicKeyPKIDPairEntryByBLSPublicKeyWithTxn(txn, snap, blsPublicKey, snapshotAtEpochNumber)
+		return innerErr
+	})
+	return ret, err
+}
+
+func DBGetSnapshotValidatorBLSPublicKeyPKIDPairEntryByBLSPublicKeyWithTxn(txn *badger.Txn, snap *Snapshot, blsPublicKey *bls.PublicKey, snapshotAtEpochNumber uint64) (*BLSPublicKeyPKIDPairEntry, error) {
```
Comment (tholonious) [inline lib/pos_snapshot_entries.go:681]:
```suggestion
func DBGetSnapshotValidatorBLSPublicKeyPKIDPairEntryWithTxn(txn *badger.Txn, snap *Snapshot, blsPublicKey *bls.PublicKey, snapshotAtEpochNumber uint64) (*BLSPublicKeyPKIDPairEntry, error) {
```

File: lib/pos_snapshot_entries.go [hunk 3]
```diff
@@ -621,6 +624,184 @@ func DBEnumerateAllCurrentValidators(handle *badger.DB, pkidsToSkip []*PKID) ([]
...
+			return nil, nil
+		}
+		return nil, errors.Wrapf(err, "DBGetSnapshotValidatorBLSPublicKeyPKIDPairEntryByBLSPublicKeyWithTxn: problem retrieving BLSPublicKeyPKIDPairEntry")
+	}
+
+	// Decode from bytes.
+	blsPublicKeyPKIDPairEntry := &BLSPublicKeyPKIDPairEntry{}
+	rr := bytes.NewReader(blsPublicKeyPKIDPairEntryBytes)
+	if exist, err := DecodeFromBytes(blsPublicKeyPKIDPairEntry, rr); !exist || err != nil {
+		return nil, errors.Wrapf(err, "DBGetSnapshotValidatorBLSPublicKeyPKIDPairEntryByBLSPublicKeyWithTxn: problem decoding BLSPublicKeyPKIDPairEntry")
```
Comment (tholonious) [inline lib/pos_snapshot_entries.go:697]:
```suggestion
		return nil, errors.Wrapf(err, "DBGetSnapshotValidatorBLSPublicKeyPKIDPairEntryWithTxn: problem decoding BLSPublicKeyPKIDPairEntry")
```

File: lib/pos_snapshot_entries.go [hunk 4]
```diff
@@ -621,6 +624,184 @@ func DBEnumerateAllCurrentValidators(handle *badger.DB, pkidsToSkip []*PKID) ([]
...
+func DBGetSnapshotValidatorBLSPublicKeyPKIDPairEntryByBLSPublicKeyWithTxn(txn *badger.Txn, snap *Snapshot, blsPublicKey *bls.PublicKey, snapshotAtEpochNumber uint64) (*BLSPublicKeyPKIDPairEntry, error) {
+	// Retrieve from db.
+	key := DBKeyForSnapshotValidatorBLSPublicKeyPKIDPairEntry(&BLSPublicKeyPKIDPairEntry{BLSPublicKey: blsPublicKey}, snapshotAtEpochNumber)
+	blsPublicKeyPKIDPairEntryBytes, err := DBGetWithTxn(txn, snap, key)
+	if err != nil {
+		// We don't want to error if the key isn't found. Instead, return nil.
+		if errors.Is(err, badger.ErrKeyNotFound) {
+			return nil, nil
+		}
+		return nil, errors.Wrapf(err, "DBGetSnapshotValidatorBLSPublicKeyPKIDPairEntryByBLSPublicKeyWithTxn: problem retrieving BLSPublicKeyPKIDPairEntry")
```
Comment (tholonious) [inline lib/pos_snapshot_entries.go:690]:
```suggestion
		return nil, errors.Wrapf(err, "DBGetSnapshotValidatorBLSPublicKeyPKIDPairEntryWithTxn: problem retrieving BLSPublicKeyPKIDPairEntry")
```

File: lib/pos_snapshot_entries.go [hunk 5]
```diff
@@ -621,6 +624,184 @@ func DBEnumerateAllCurrentValidators(handle *badger.DB, pkidsToSkip []*PKID) ([]
...
+	key = append(key, EncodeUint64(snapshotAtEpochNumber)...)
+	key = append(key, blsPublicKeyPKIDPairEntry.BLSPublicKey.ToBytes()...)
+	return key
+}
+
+func DBGetSnapshotValidatorBLSPublicKeyPKIDPairEntryByBLSPublicKey(handle *badger.DB, snap *Snapshot, blsPublicKey *bls.PublicKey, snapshotAtEpochNumber uint64) (*BLSPublicKeyPKIDPairEntry, error) {
+	var ret *BLSPublicKeyPKIDPairEntry
+	err := handle.View(func(txn *badger.Txn) error {
+		var innerErr error
+		ret, innerErr = DBGetSnapshotValidatorBLSPublicKeyPKIDPairEntryByBLSPublicKeyWithTxn(txn, snap, blsPublicKey, snapshotAtEpochNumber)
```
Comment (tholonious) [inline lib/pos_snapshot_entries.go:675]:
```suggestion
		ret, innerErr = DBGetSnapshotValidatorBLSPublicKeyPKIDPairEntryWithTxn(txn, snap, blsPublicKey, snapshotAtEpochNumber)
```

File: lib/pos_snapshot_entries.go [hunk 6]
```diff
@@ -621,6 +624,184 @@ func DBEnumerateAllCurrentValidators(handle *badger.DB, pkidsToSkip []*PKID) ([]
...
+		return errors.Wrap(
+			err, "DBDeleteSnapshotValidatorBLSPublicKeyPKIDPairEntryWithTxn: problem deleting BLSPublicKeyPKIDPairEntry from index PrefixSnapshotValidatorBLSPublicKeyPKIDPairEntry",
+		)
+	}
+	return nil
+}
+
+func (bav *UtxoView) _setSnapshotValidatorBLSPublicKeyPKIDPairEntry(blsPublicKeyPKIDPairEntry *BLSPublicKeyPKIDPairEntry, snapshotAtEpochNumber uint64) {
+	if blsPublicKeyPKIDPairEntry == nil {
+		glog.Errorf("_setValidatorBLSPublicKeyEntry: called with nil entry, this should never happen")
```
Comment (tholonious) [inline lib/pos_snapshot_entries.go:752]:
```suggestion
		glog.Errorf("_setSnapshotValidatorBLSPublicKeyPKIDPairEntry: called with nil entry, this should never happen")
```

**Review by tholonious — APPROVED** [review]
Just naming nits

**Review by tholonious — APPROVED** [review]

**Pull request merged**